[router] refactor dynamic mcp server

key4ng opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

When using dynamic MCP servers within a request, several issues exist in the current implementation:

- A user can inadvertently access another user’s dynamic MCP server stored in the shared tool inventory.
- A request can access MCP servers that were registered in **previous** requests, leading to cross-request leakage.
- If multiple users define tools with the same name but different server labels, the system stores only one entry, causing unintended overrides.
- The current design only supports **one** MCP server per request, limiting flexibility.

## Modifications

- Introduced a **per-request tool map**, which exists only for the lifetime of the request.  
  - Uses composite keys `(server_label, tool_name)`.  
  - Values contain both the `Tool` and its MCP server `url`.  
  - The `url` is used to get or reuse a client from the connection pool.
- Refactored function-call naming from `<tool_name>` to `<server_label>__<tool_name>`, enabling correct routing to the associated MCP server during invocation.
- Applied these changes across the OpenAI, gRPC, and Harmony routers.

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

key4ng committed:
[router] refactor dynamic mcp server

key4ng committed:
fix import

CatherineSue commented:
> Refactored function-call naming from <tool_name> to <server_label>__<tool_name>

Is this function call or mcp call?

> Introduced a per-request tool map, which exists only for the lifetime of the request.
Uses composite keys (server_label, tool_name).
Values contain both the Tool and its MCP server url.
The url is used to get or reuse a client from the connection pool.

Can we have a more detailed section regarding the data flow? It is a bit hard to grasp at the moment.

CatherineSue commented on sgl-router/src/mcp/manager.rs:
```diff
@@ -161,6 +158,54 @@ impl McpManager {
+        &self,
+        dynamic_servers: &[(String, String)],
+    ) -> std::collections::HashMap<(String, String), (Tool, String)> {
+        use std::collections::HashMap;
```
please use imports at the top

CatherineSue commented on sgl-router/src/routers/grpc/harmony/responses.rs:
```diff
@@ -1418,6 +1467,38 @@ pub fn convert_mcp_tools_to_response_tools(mcp_tools: &[mcp::Tool]) -> Vec<Respo
+///
+/// This function formats tool names as `server_label__tool_name` to prevent name collisions
+/// across multiple MCP servers and enable server-specific tool routing.
+fn convert_mcp_tools_map_to_response_tools(
```
where is `convert_mcp_tools_to_response_tools` used then? why do we need two of these now?

key4ng commented on sgl-router/src/routers/grpc/common/responses/utils.rs:
```diff
@@ -9,6 +9,8 @@ use axum::{
 use tracing::{debug, error, warn};
 
+// Re-export MCP utilities for use by grpc routers
+pub use crate::routers::openai::mcp::extract_dynamic_mcp_servers;
```
good point. I'm assuming this method should be in a common place that openai/grpc/harmony can also used, but now since this refactor was not done yet and I don't want the files in grpc/harmony has direct dependency in openai router, so temporarily put it here. I can add a TODO here

key4ng committed:
resolve comments

key4ng commented on sgl-router/src/routers/grpc/harmony/responses.rs:
```diff
@@ -1150,12 +1198,31 @@ async fn execute_mcp_tools(
+        let (server_label, original_tool_name) =
+            if let Some(sep_pos) = formatted_tool_name.find("__") {
+                let server_label = &formatted_tool_name[..sep_pos];
+                let tool_name = &formatted_tool_name[sep_pos + 2..]; // +2 to skip "__"
```
yess..changed it to use regex

key4ng committed:
add todo for reimport

key4ng committed:
refactoring inventory

key4ng committed:
add streaming change

key4ng committed:
improve the common module import

key4ng committed:
move method to common

key4ng committed:
clean code

key4ng committed:
clean useless comments

key4ng committed:
fix fmt

key4ng commented:
@slin1237 @CatherineSue thanks for the comments, pls have another look.
some key changes since last review:
- Tool Inventory:
    - The DashMap definition changed to composite key(server_label, tool_name) and value(Tool, server_url)
    - cleaned no used prompt and resource related code
    - Instead of creating a new hashMap in my initial implementation, dynamic mcp and static mcp will both use tool Inventory. The dynamic MCP tool inventory is scoped to a single request, while the static MCP tool inventory persists for the lifetime of the server.
- mcp client: static mcp client used server name as a key for cache while dynamic client used server url. Changed static mcp client to use server url as well for consistency
- created a common.rs under `/router` for the codes used by both grpc and openai routers

some data flow graphs:
**Static Servers (from config)**
```
Startup
  ↓
McpManager::new(config)
  ↓
For each server in config:
  ├─ Compute server_url = server_key(config)  // URL from transport
  ├─ Connect to server
  ├─ List tools from server
  ├─ Store in static_clients[server_url] = client
  └─ Insert tools into inventory[(server_url, tool_name)] = (tool, server_url)
  ↓
Manager ready with static inventory
```

**Dynamic Servers (from requests)**

```
Request with MCP tools
  ↓
extract_dynamic_mcp_servers(tools)  // routers/common.rs
  ├─ Filters tools by type == ResponseToolType::Mcp
  ├─ Validates URLs (http/https only)
  ├─ Generates server_label (from tool.server_label or auto-generated)
  └─ Returns Vec<(server_label, server_url)>
  ↓
ensure_request_mcp_client(mcp_manager, tools)  // routers/common.rs
  ├─ Creates McpServerConfig from first MCP tool
  └─ Calls mcp_manager.get_or_create_client(config)
      ├─ Check static_clients[server_url]
      ├─ If not found, use connection_pool.get_or_create(server_url)
      │   ├─ Check cache (max 200 connections)
      │   ├─ If cache miss, connect and add to pool
      │   └─ LRU eviction when pool is full
      └─ Return client
  ↓
build_request_inventory(dynamic_servers)
  ├─ Create new ToolInventory
  ├─ Merge static tools from manager.inventory
  └─ For each (server_label, server_url):
      ├─ Get client for server_url
      ├─ List tools from client
      └─ Insert into request_inventory[(server_label, tool_name)] = (tool, server_url)
  ↓
Per-request inventory with static + dynamic tools
```

**Tool Execution Flow**

```
Tool call request (formatted as "server_label__tool_name")
  ↓
parse_tool_name(formatted_name)  // routers/common.rs
  ├─ Uses regex: ^([^_]+(?:_[^_]+)*)__(.+)$
  └─ Returns (server_label, tool_name)
  ↓
call_tool_from_inventory(inventory, server_label, tool_name, args)
  ├─ Look up tool: inventory.get_tool(server_label, tool_name)
  │   └─ Returns (Tool, server_url)
  ├─ Get client: manager.get_client(server_url)
  │   ├─ Check static_clients[server_url]
  │   └─ Fallback to connection_pool.get(server_url)
  ├─ Convert args using tool schema (type coercion)
  └─ Execute: client.call_tool(tool_name, args)
      └─ Returns CallToolResult
```
**Integration with router**
```
gRPC Responses API
  ↓
Regular Router
  ├─ extract_dynamic_mcp_servers(tools) → dynamic_servers
  ├─ build_request_inventory(dynamic_servers) → inventory
  └─ For each iteration:
      ├─ prepare_chat_tools_and_choice(request, inventory)
      │   └─ Merges function + MCP tools with formatted names
      ├─ Send to backend
      └─ execute_mcp_tools(mcp_manager, tool_calls, inventory)
          └─ For each tool call:
              ├─ parse_tool_name(formatted_name)
              └─ call_tool_from_inventory(inventory, server_label, tool_name, args)
```

key4ng committed:
resolve comments

key4ng committed:
using string instead of composite key

key4ng committed:
mcp encapsulation

key4ng committed:
clean comments

key4ng committed:
fix tool name in response for harmony
